core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala

```diff
@@ -276,6 +289,165 @@ class ConsumerBounceTest extends BaseRequestTest with Logging {
+    * the group should be forced to rebalance when it becomes hosted on a Coordinator with the new config.
+    * Then, 1 consumer should be left out of the group.
+    */
+  @Test
+  def testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup(): Unit = {
+    val topic = "group-max-size-test"
+    val maxGroupSize = 2
+    val consumerCount = maxGroupSize + 1
+    var recordsProduced = maxGroupSize * 100
+    val partitionCount = consumerCount * 2
+    if (recordsProduced % partitionCount != 0) {
+      // ensure even record distribution per partition
+      recordsProduced += partitionCount - recordsProduced % partitionCount
+    }
+    val executor = Executors.newScheduledThreadPool(consumerCount * 2)
+    this.consumerConfig.setProperty(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, "60000")
+    this.consumerConfig.setProperty(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG, "1000")
+    this.consumerConfig.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false")
+    val producer = createProducer()
+    createTopic(topic, numPartitions = partitionCount, replicationFactor = numBrokers)
+    val stableConsumers = createConsumersWithGroupId("group2", consumerCount, executor, topic = topic)
+
+    // assert group is stable and working
+    sendRecords(producer, recordsProduced, topic, numPartitions = Some(partitionCount))
+    stableConsumers.foreach(cons => {
+      receiveAtLeastRecords(cons, recordsProduced / consumerCount, 10000)
+    })
+
+    // roll all brokers with a lesser max group size to make sure coordinator has the new config
+    val newConfigs = generateKafkaConfigs(maxGroupSize.toString)
+    val receivedExceptions = new ArrayBuffer[Throwable]()
+    var kickedOutConsumerIdx: Option[Int] = None
+    val lock = new ReentrantLock
+    // restart brokers until the group moves to a Coordinator with the new config
+    breakable { for (broker <- servers.indices) {
+      killBroker(broker)
+      sendRecords(producer, recordsProduced, topic, numPartitions = Some(partitionCount))
+
+      var successfulConsumes = 0
+
+      // compute consumptions in a non-blocking way in order to account for the rebalance once the group.size takes effect
+      val consumeFutures = new ArrayBuffer[SFuture[Any]]
+      implicit val executorContext: ExecutionContextExecutor = ExecutionContext.fromExecutor(executor)
+      stableConsumers.indices.foreach(idx => {
+        val currentConsumer = stableConsumers(idx)
+        val consumeFuture = SFuture {
+          try {
+            receiveAtLeastRecords(currentConsumer, recordsProduced / consumerCount, 10000)
+            CoreUtils.inLock(lock) { successfulConsumes += 1 }
+          } catch {
+            case e: Throwable =>
+              if (e.isInstanceOf[GroupMaxSizeReachedException]) {
+                kickedOutConsumerIdx = Some(idx)
+              }
+
+              CoreUtils.inLock(lock) { receivedExceptions += e }
```

Comment:
Seems we might be able to simplify this. If we receive anything other than `GroupMaxSizeReachedException`, couldn't we fail directly? Then it could be a simple `AtomicBoolean` tracking whether or not we have received exactly one exception.